File: ipadOS/RCForb/Sources/Networking/UDPClient.swift
import Foundation
import Network

/// V10 UDP client ported from udp-client.ts
class UDPClient {
    private var connection: NWConnection?
    private var isConnectedFlag = false
    private var isDataFlowingFlag = false
    private var lastServerData = Date()
    private var heartbeatTimer: DispatchSourceTimer?
    private var pingTimer: DispatchSourceTimer?
    private var timeoutTimer: DispatchSourceTimer?
    private let queue = DispatchQueue(label: "com.rcforb.udp", qos: .userInteractive)

    var onAudio: ((Data) -> Void)?
    var onCommand: ((String) -> Void)?
    var onControl: ((UInt8) -> Void)?
    var onDisconnected: (() -> Void)?

    private var onDataFlowResume: ((Bool) -> Void)?

    var isConnected: Bool { isConnectedFlag }
    var isDataFlowing: Bool { isDataFlowingFlag }

    func connect(host: String, port: Int) async -> Bool {
        let endpoint = NWEndpoint.hostPort(host: NWEndpoint.Host(host), port: NWEndpoint.Port(integerLiteral: UInt16(port)))
        let params = NWParameters.udp
        params.requiredInterfaceType = .other
        let conn = NWConnection(to: endpoint, using: params)
        self.connection = conn

        return await withCheckedContinuation { continuation in
            var resumed = false
            let resume: (Bool) -> Void = { value in
                guard !resumed else { return }
                resumed = true
                continuation.resume(returning: value)
            }

            conn.stateUpdateHandler = { [weak self] state in
                switch state {
                case .ready:
                    self?.isConnectedFlag = true
                    self?.isDataFlowingFlag = false
                    self?.lastServerData = Date()

                    let userInData = Data([ControlByte.USERIN])
                    for _ in 0..<3 {
                        self?.sendRaw(userInData)
                    }

                    self?.startHeartbeat()
                    self?.startReceiving()
                    resume(true)
                case .failed, .cancelled:
                    self?.handleDisconnect()
                    resume(false)
                default:
                    break
                }
            }
            conn.start(queue: queue)

            queue.asyncAfter(deadline: .now() + 10) { [weak self] in
                guard let self, !self.isConnectedFlag else { return }
                conn.cancel()
                resume(false)
            }
        }
    }

    func disconnect() {
        guard isConnectedFlag || connection != nil else { return }
        sendRaw(Data([ControlByte.USEROUT]))
        stopTimers()
        isConnectedFlag = false
        isDataFlowingFlag = false
        connection?.cancel()
        connection = nil
        onDisconnected?()
    }

    private func handleDisconnect() {
        stopTimers()
        isConnectedFlag = false
        isDataFlowingFlag = false
        connection?.cancel()
        connection = nil
        onDisconnected?()
    }

    func sendRaw(_ data: Data) {
        connection?.send(content: data, completion: .contentProcessed { _ in })
    }

    func sendCommandString(_ text: String) {
        var packet = Data([ControlByte.UTF8STRING])
        packet.append(text.data(using: .utf8) ?? Data())
        sendRaw(packet)
    }

    func sendPTT(_ on: Bool) {
        sendRaw(Data([on ? ControlByte.PTT : ControlByte.PTT_OFF]))
    }

    func sendAudio(_ data: Data) {
        sendRaw(data)
    }

    private func startReceiving() {
        connection?.receiveMessage { [weak self] data, _, _, error in
            guard let self, let data, error == nil else { return }
            self.processPacket(data)
            self.startReceiving()
        }
    }

    private func processPacket(_ data: Data) {
        guard !data.isEmpty else { return }

        lastServerData = Date()
        if !isDataFlowingFlag {
            isDataFlowingFlag = true
            onDataFlowResume?(true)
            onDataFlowResume = nil
        }

        let type = classifyPacket(data)
        switch type {
        case .audio:
            onAudio?(data)
        case .command:
            let text = String(data: data.dropFirst(), encoding: .utf8) ?? ""
            onCommand?(text)
        case .pttOn:
            onControl?(ControlByte.PTT)
        case .pttOff:
            onControl?(ControlByte.PTT_OFF)
        case .keyOn:
            onControl?(ControlByte.KEY_ON)
        case .keyOff:
            onControl?(ControlByte.KEY_OFF)
        case .userOut:
            handleDisconnect()
        case .heartbeat, .pingpong, .userIn:
            break
        }
    }

    private func startHeartbeat() {
        heartbeatTimer = DispatchSource.makeTimerSource(queue: queue)
        heartbeatTimer?.schedule(deadline: .now(), repeating: .milliseconds(HEARTBEAT_INTERVAL_MS))
        heartbeatTimer?.setEventHandler { [weak self] in
            self?.sendRaw(Data([ControlByte.HEARTBEAT]))
        }
        heartbeatTimer?.resume()

        pingTimer = DispatchSource.makeTimerSource(queue: queue)
        pingTimer?.schedule(deadline: .now(), repeating: .milliseconds(PING_INTERVAL_MS))
        pingTimer?.setEventHandler { [weak self] in
            self?.sendRaw(Data([ControlByte.PINGPONG]))
        }
        pingTimer?.resume()

        timeoutTimer = DispatchSource.makeTimerSource(queue: queue)
        timeoutTimer?.schedule(deadline: .now() + 1, repeating: 1)
        timeoutTimer?.setEventHandler { [weak self] in
            guard let self else { return }
            if Date().timeIntervalSince(self.lastServerData) * 1000 > Double(HEARTBEAT_TIMEOUT_MS) {
                print("[UDPClient] Heartbeat timeout, disconnecting")
                self.handleDisconnect()
            }
        }
        timeoutTimer?.resume()
    }

    private func stopTimers() {
        heartbeatTimer?.cancel(); heartbeatTimer = nil
        pingTimer?.cancel(); pingTimer = nil
        timeoutTimer?.cancel(); timeoutTimer = nil
    }

    func waitForDataFlow(timeoutMs: Int = 8000) async -> Bool {
        if isDataFlowingFlag { return true }
        return await withCheckedContinuation { continuation in
            var resumed = false
            let resume: (Bool) -> Void = { value in
                guard !resumed else { return }
                resumed = true
                continuation.resume(returning: value)
            }
            self.onDataFlowResume = resume
            queue.asyncAfter(deadline: .now() + .milliseconds(timeoutMs)) { [weak self] in
                self?.onDataFlowResume?(false)
                self?.onDataFlowResume = nil
            }
        }
    }
}

File: android/app/src/main/java/com/rcforb/android/networking/UDPClient.kt
package com.rcforb.android.networking

import android.util.Log
import com.rcforb.android.protocol.*
import kotlinx.coroutines.*
import java.net.DatagramPacket
import java.net.DatagramSocket
import java.net.InetAddress
import java.net.SocketTimeoutException

class UDPClient {
    private var socket: DatagramSocket? = null
    private var address: InetAddress? = null
    private var port: Int = 0
    private var isConnectedFlag = false
    private var isDataFlowingFlag = false
    private var lastServerData = System.currentTimeMillis()
    private var receiveJob: Job? = null
    private var heartbeatJob: Job? = null
    private var pingJob: Job? = null
    private var timeoutJob: Job? = null
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    var onAudio: ((ByteArray) -> Unit)? = null
    var onCommand: ((String) -> Unit)? = null
    var onControl: ((Byte) -> Unit)? = null
    var onDisconnected: (() -> Unit)? = null

    private var dataFlowContinuation: CancellableContinuation<Boolean>? = null

    val isConnected: Boolean get() = isConnectedFlag
    val isDataFlowing: Boolean get() = isDataFlowingFlag

    suspend fun connect(host: String, port: Int): Boolean = withContext(Dispatchers.IO) {
        try {
            val addr = InetAddress.getByName(host)
            val sock = DatagramSocket()
            sock.soTimeout = 500
            sock.connect(addr, port)

            this@UDPClient.socket = sock
            this@UDPClient.address = addr
            this@UDPClient.port = port
            isConnectedFlag = true
            isDataFlowingFlag = false
            lastServerData = System.currentTimeMillis()

            // Send USERIN x3
            repeat(3) { sendRaw(byteArrayOf(ControlByte.USERIN)) }

            startHeartbeat()
            startReceiving()
            true
        } catch (e: Exception) {
            Log.e("UDPClient", "Connect failed", e)
            false
        }
    }

    fun disconnect() {
        if (!isConnectedFlag && socket == null) return
        sendRaw(byteArrayOf(ControlByte.USEROUT))
        stopAll()
        onDisconnected?.invoke()
    }

    private fun handleDisconnect() {
        stopAll()
        onDisconnected?.invoke()
    }

    private fun stopAll() {
        isConnectedFlag = false
        isDataFlowingFlag = false
        receiveJob?.cancel()
        heartbeatJob?.cancel()
        pingJob?.cancel()
        timeoutJob?.cancel()
        try { socket?.close() } catch (_: Exception) {}
        socket = null
    }

    fun sendRaw(data: ByteArray) {
        val sock = socket ?: return
        val addr = address ?: return
        try {
            val packet = DatagramPacket(data, data.size, addr, port)
            sock.send(packet)
        } catch (_: Exception) {}
    }

    fun sendCommandString(text: String) {
        val textBytes = text.toByteArray(Charsets.UTF_8)
        val packet = ByteArray(1 + textBytes.size)
        packet[0] = ControlByte.UTF8STRING
        System.arraycopy(textBytes, 0, packet, 1, textBytes.size)
        sendRaw(packet)
    }

    fun sendPTT(on: Boolean) {
        sendRaw(byteArrayOf(if (on) ControlByte.PTT else ControlByte.PTT_OFF))
    }

    fun sendAudio(data: ByteArray) {
        sendRaw(data)
    }

    private fun startReceiving() {
        receiveJob = scope.launch {
            val buf = ByteArray(65536)
            while (isActive && isConnectedFlag) {
                try {
                    val packet = DatagramPacket(buf, buf.size)
                    socket?.receive(packet)
                    if (packet.length > 0) {
                        val data = buf.copyOfRange(0, packet.length)
                        processPacket(data)
                    }
                } catch (_: SocketTimeoutException) {
                    // Normal — just retry
                } catch (e: Exception) {
                    if (isConnectedFlag) {
                        Log.e("UDPClient", "Receive error", e)
                        handleDisconnect()
                    }
                    break
                }
            }
        }
    }

    private fun processPacket(data: ByteArray) {
        if (data.isEmpty()) return

        lastServerData = System.currentTimeMillis()
        if (!isDataFlowingFlag) {
            isDataFlowingFlag = true
            dataFlowContinuation?.resumeWith(Result.success(true))
            dataFlowContinuation = null
        }

        when (classifyPacket(data)) {
            PacketType.AUDIO -> onAudio?.invoke(data)
            PacketType.COMMAND -> {
                val text = String(data, 1, data.size - 1, Charsets.UTF_8)
                onCommand?.invoke(text)
            }
            PacketType.PTT_ON -> onControl?.invoke(ControlByte.PTT)
            PacketType.PTT_OFF -> onControl?.invoke(ControlByte.PTT_OFF)
            PacketType.KEY_ON -> onControl?.invoke(ControlByte.KEY_ON)
            PacketType.KEY_OFF -> onControl?.invoke(ControlByte.KEY_OFF)
            PacketType.USER_OUT -> handleDisconnect()
            PacketType.HEARTBEAT, PacketType.PINGPONG, PacketType.USER_IN -> {}
        }
    }

    private fun startHeartbeat() {
        heartbeatJob = scope.launch {
            while (isActive && isConnectedFlag) {
                sendRaw(byteArrayOf(ControlByte.HEARTBEAT))
                delay(HEARTBEAT_INTERVAL_MS)
            }
        }
        pingJob = scope.launch {
            while (isActive && isConnectedFlag) {
                sendRaw(byteArrayOf(ControlByte.PINGPONG))
                delay(PING_INTERVAL_MS)
            }
        }
        timeoutJob = scope.launch {
            while (isActive && isConnectedFlag) {
                delay(1000)
                if (System.currentTimeMillis() - lastServerData > HEARTBEAT_TIMEOUT_MS) {
                    Log.w("UDPClient", "Heartbeat timeout, disconnecting")
                    handleDisconnect()
                    break
                }
            }
        }
    }

    suspend fun waitForDataFlow(timeoutMs: Long = 8000): Boolean {
        if (isDataFlowingFlag) return true
        return withTimeoutOrNull(timeoutMs) {
            suspendCancellableCoroutine { cont ->
                dataFlowContinuation = cont
            }
        } ?: false
    }
}
